### PR TITLE
Add ability to restrict namespaces being watched.

### PR DESCRIPTION
**Pull request by cfchase**
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] JIRA link(s): [RHODS-2732](https://issues.redhat.com/browse/RHODS-2732)
- [x] The Jira story is acked
- [x] An entry has been added to the latest build document in [Build Announcements Folder](https://drive.google.com/drive/folders/1sgkK1WZgGo9CXsLizNe0GbAzVKuSKrZL).
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious)

**Testing:**
 1. Install rhods using https://gitlab.cee.redhat.com/data-hub/olminstall with catalogsource: `quay.io/modh/rhods-operator-live-catalog:1.7.0-rhods-2732` [from quay/modh/rhods-operator-live-catalog](https://quay.io/repository/modh/rhods-operator-live-catalog?tag=1.7.0-rhods-2732&tab=tags)
2. Observe operator memory usage redhat-ods-operator using OpenShift "observe" dashboards
3. Make config changes in `redhat-ods-applications`, `redhat-ods-operator` and make sure the operator reacts appropriately (e.g. change replicas of deployment to 2 and watch it get reset by the operator).
4. Create 1000+ namespaces
```bash
for i in {1..1000}
do
   oc create ns dummyns${i}
done
```
Memory should not increase with namespaces
5. Uninstall rhods using the proper `managed-odh` config map with labels `api.openshift.com/addon-managed-odh-delete: 'true'`

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -129,8 +140,9 @@ func main() {
 
 	// CreateServiceMonitors will automatically create the prometheus-operator ServiceMonitor resources
 	// necessary to configure Prometheus to scrape metrics from this operator.
+	operatorNamespace, _ := k8sutil.GetOperatorNamespace()
```
Comment (cfchase) [inline cmd/manager/main.go:143]: Previously, this had been set to the `WATCH_NAMESPACE` which was always `""`.  That resulted in:
```
time="2022-02-03T01:23:03Z" level=error msg="Could not create ServiceMonitor object. Error: an empty namespace may not be set during creation."
```

We could just keep that behavior by setting the `namespace` to `""`, but if we might as well comment it out.  If we want to enable this for the operator, we should turn it on for the namespace the operator is in, `k8sutil.GetOperatorNamespace()`

File: cmd/manager/main.go [hunk 2]
```diff
@@ -65,10 +66,11 @@ func main() {
 
 	printVersion()
 
-	namespace, err := k8sutil.GetWatchNamespace()
+	watchNamespace, err := k8sutil.GetWatchNamespace()
```
Comment (cfchase) [inline cmd/manager/main.go:69]: This comes from the env var `WATCH_NAMESPACE`.  That is set previously this was set to `metadata.annotations.olm.targetNamespaces` which was `""` and then ignored.  The operator was always using `""` which watches all namespaces.

`WATCH_NAMESPACE`:
- Leaving it as `""` will continue with the current behavior of monitoring all namespaces.
- Setting it to a single namespace will monitor that namespace.
- Setting it to comma separated namespaces in the CSV, e.g. `redhat-ods-operator,redhat-ods-applications,redhat-ods-monitoring,rhods-notebooks`., will monitor those namespaces.

**Comment by cfchase** [comment]
@LaVLaS Can you let me know if you see a CI issue while you're reviewing?

**Comment by VaishnaviHire** [comment]
@cfchase I tested the given live build on PSI cluster. I think we are accessing some cluster-wide resources, specifically https://github.com/red-hat-data-services/opendatahub-operator/blob/master/pkg/controller/kfdef/kfdef_controller.go#L613, limiting the scope might throw errors causing the controller to reconcile during uninstall.

**Comment by anishasthana** [comment]
The functionality is there, but I'm noticing weird behavior during uninstall. The redhat-ods-operator namespace (and contents) is sticking around... somehow

**Comment by VaishnaviHire** [comment]
> The functionality is there, but I'm noticing weird behavior during uninstall. The redhat-ods-operator namespace (and contents) is sticking around... somehow.
> 
> This is on OSD.

The probable cause for this might be that we are limiting the permissions for deleting the csv. As part of last step of uninstallation, we [remove](https://github.com/red-hat-data-services/opendatahub-operator/blob/master/pkg/controller/kfdef/kfdef_controller.go#L668) csv that is a requirement for [addon deletion](https://gitlab.cee.redhat.com/service/managed-tenants/-/blob/main/docs/ocm/addons_external_resources.md#addon-deletion).

**Comment by cfchase** [comment]
yeah, not looking good for this approach.  Will have to look a different method.  Unless I figure something out, I'll probably close this PR

**Comment by cfchase** [comment]
> @cfchase I tested the given live build on PSI cluster. I think we are accessing some cluster-wide resources, specifically https://github.com/red-hat-data-services/opendatahub-operator/blob/master/pkg/controller/kfdef/kfdef_controller.go#L613, limiting the scope might throw errors causing the controller to reconcile during uninstall.

Turns out this was the root cause, the CSV permissions worked.  That means the overall solution should work if we fix this.

**Comment by cfchase** [comment]
Updated and retested on OSIA and OSD

@VaishnaviHire @LaVLaS @crobby ready for re-review
@LaVLaS anything I can do about the CI failures?
@anishasthana updated live build and ready for retest

**Comment by crobby** [comment]
> > @cfchase I tested the given live build on PSI cluster. I think we are accessing some cluster-wide resources, specifically https://github.com/red-hat-data-services/opendatahub-operator/blob/master/pkg/controller/kfdef/kfdef_controller.go#L613, limiting the scope might throw errors causing the controller to reconcile during uninstall.
> 
> Turns out this was the root cause, the CSV permissions worked. That means the overall solution should work if we fix this.

Looking at the last code update.  We just stripped out the part that dealt with certmanager?  Was it unnecessary code to start with?

**Review by crobby — APPROVED** [review]
I've tested this a few times now.  Seems to work well.
/lgtm

**Pull request merged**